src/servers/src/elasticsearch.rs

```diff
@@ -0,0 +1,366 @@
+    HeaderMap::from_iter([
+        (
+            axum::http::header::CONTENT_TYPE,
+            HeaderValue::from_static("application/json"),
+        ),
+        // Logstash needs this header to identify the product.
+        (
+            HeaderName::from_static("x-elastic-product"),
+            HeaderValue::from_static("Elasticsearch"),
+        ),
+    ])
+}
+
+// The input will be Elasticsearch bulk request in NDJSON format.
+// For example, the input will be like this:
+// { "index" : { "_index" : "test", "_id" : "1" } }
+// { "field1" : "value1" }
+// { "index" : { "_index" : "test", "_id" : "2" } }
+// { "field2" : "value2" }
+fn convert_es_input_to_log_values(
+    input: &str,
+    msg_field: &Option<String>,
+) -> ServersResult<Vec<Value>> {
+    // Read the ndjson payload and convert it to `Vec<Value>`. Return error if the input is not a valid JSON.
+    let values: Vec<Value> = Deserializer::from_str(input)
+        .into_iter::<Value>()
+        .collect::<Result<_, _>>()
+        .context(ParseJsonSnafu)?;
+
+    // Check if the input is empty.
+    ensure!(
+        !values.is_empty(),
+        InvalidElasticsearchInputSnafu {
+            reason: "empty bulk request".to_string(),
+        }
+    );
+
+    let mut log_values: Vec<Value> = Vec::new();
+
+    // For Elasticsearch post `_bulk` API, each chunk contains two objects:
+    // 1. The first object is the command, it should be `create` or `index`. `create` is used for insert, `index` is used for upsert.
+    // 2. The second object is the document data.
+    let mut is_document = false;
+    for v in values {
+        if !is_document {
+            // Read the first object to get the command, it should be `create` or `index`.
+            ensure!(
+                v.get("create").is_some() || v.get("index").is_some(),
+                InvalidElasticsearchInputSnafu {
+                    reason: format!(
+                        "invalid bulk request, expected 'create' or 'index' but got {:?}",
+                        v
+                    ),
+                }
+            );
+            is_document = true;
+            continue;
+        }
+
+        // It means the second object is the document data.
+        if is_document {
+            // If the msg_field is provided, fetch the value of the field from the document data.
+            if let Some(msg_field) = msg_field {
+                if let Some(Value::String(message)) = v.get(msg_field) {
+                    let value = match serde_json::from_str::<Value>(message) {
+                        Ok(value) => value,
+                        // If the message is not a valid JSON, just use the original message as the log value.
+                        Err(_) => Value::String(message.to_string()),
+                    };
+                    log_values.push(value);
+                }
+            } else {
+                log_values.push(v);
+            }
+            is_document = false;
+        }
```

Comment:
If `is_document=false`, do we need to print a warning message? Otherwise, it would be hard to debug such requests.

Reply:
`is_document` is a flag variable that indicates that the current object is a document. I add some comments for its usage.